backend/src/chat.gateway.ts

```diff
@@ -0,0 +1,170 @@
+  async handleLeave(client: Socket, payload: ChatJoinDto) {
+    this.logger.log(`${client.id} wants to leave channel [${payload.channel}]`);
+    this.checkAuth(client);
+    // TODO: Check if User leaving affects ownership
+    client.leave(payload.channel);
+    const membershipArray = await this.membershipsService.findAll({
+      channel: payload.channel,
+      user: this.authenticatedSockets.get(client.id).id.toString(),
+    });
+    if (membershipArray && membershipArray.length === 1) {
+      const membership = membershipArray[0];
+      if (
+        membership &&
+        membership.id &&
+        membership.userId === this.authenticatedSockets.get(client.id).id &&
+        membership.channelId === parseInt(payload.channel)
+      ) {
+        await this.membershipsService.remove(membership.id);
+        return `SUCCESS: left channel and room ${payload.channel}`;
+      }
+    }
+  }
+
+  @SubscribeMessage('chat-auth')
+  async handleAuth(client: Socket, payload: any) {
+    this.logger.log(
+      `Client ${client.id} is trying to auth socket with token: ${payload.authorization}`,
+    );
+    try {
+      const token = payload.authorization;
+      const decode = jwt.verify(token, process.env.JWT_SECRET_KEY);
+      const user = await this.usersService.findOne(
+        +(decode as TokenUserDto).id,
+      );
+      if (!user) {
+        throw new WsException('Invalid Token');
+      }
+      this.authenticatedSockets.set(client.id, user);
+      return 'SUCCESS';
+    } catch (err) {
+      this.authenticatedSockets.delete(client.id);
+      throw new WsException('Not Authorized');
+    }
+  }
+
+  @SubscribeMessage('chat-send')
+  async handleSend(client: Socket, payload: ChatSendDto): Promise<string> {
+    this.checkAuth(client);
+    const target = payload.channel;
+    const message = payload.message;
+    if (!target || !message) {
+      return `ERROR: Invalid payload: ${payload}`;
+    }
+    this.logger.log(
+      `${client.id} wants to send ["${message}"] to target ${target}!`,
+    );
+    const messageDto = new CreateMessageDto();
+    messageDto.channelId = parseInt(payload.channel);
+    messageDto.content = message;
+    messageDto.senderId = this.authenticatedSockets.get(client.id).id;
+
+    // TODO: Check if User has permission to send message here
+    const messageResponse = await this.messagesService.create(messageDto);
+    const messageResponseDto: ResponseMessageDto = plainToInstance(
+      ResponseMessageDto,
+      messageResponse,
+      { excludeExtraneousValues: true },
+    );
+    this.server.to(target).emit('chat-message', messageResponseDto);
+    return 'Message Confirmed';
+  }
+
+  afterInit() {
+    this.logger.log('Initialized');
+  }
+
+  handleDisconnect(client: Socket) {
+    this.logger.log(`Client disconnected: ${client.id}`);
+    this.authenticatedSockets.delete(client.id);
+  }
+
+  async handleConnection(client: Socket) {
+    this.logger.log(`Client connected: ${client.id}`);
```

Comment:
Should a user socket not join every channel the user is a member (through the membership service)?

Reply:
I thought this would be quite easy to implement, but I'm having some trouble. So I opened issue #120 to do that separately, if that sounds good to you

Reply:
Sounds good!